feat: Integer update

novusnota opened the pull request:
Adds a new page to the Book — "Integers", which describes integer literals, serialization as `tl-b` types and more.

Closes #60
Closes #61

The page makes couple of references to the new page added there: https://github.com/tact-lang/tact-docs/pull/95

anton-trunov commented on pages/book/integers.mdx:
```diff
@@ -0,0 +1,107 @@
+
+## Notation
+
+Tact supports various ways of writing primitive values of `Int{:tact}` (integer literals). Most of the notations allow adding underscores (`_`) in-between digits, except for the certain case of decimal numeral system numbers, which start with $0,$ as well as for representations in strings (as seen in [nano-tons](#nano-tons) case)
```
We should probably mention that several underscores in a row (`4__2`) are not allowed, and also trailing underscores are not available (`42_`).

novusnota replied:
Yeah, I'll re-write related parts

anton-trunov commented on pages/book/integers.mdx:
```diff
@@ -0,0 +1,107 @@
+Most common and most used way of representing numbers, using the [decimal numeral system](https://en.wikipedia.org/wiki/Decimal): $123456789.$\
+You can use underscores (`_`) to improve readability: $123\_456\_789.$
+
+Alternatively, you can prefix the number with one $0$, which prohibits use of underscores and only allows decimal digits: $0123 = 123.$
```
We should mention that leading zeros are discouraged because, for instance, in TypeScript those are octal integer literals and hence there is a risk of mistakes when copying integer literals from Tact source code into TypeScript-based tests

novusnota replied:
Absolutely. Although, we could've just not introduce them to the grammar in the first place :)

anton-trunov replied:
It was already supported, before https://github.com/tact-lang/tact/pull/99 we had `integerLiteralDec = digit+`

anton-trunov commented on pages/book/integers.mdx:
```diff
@@ -0,0 +1,107 @@
+
+All runtime calculations with numbers are done at 257-bits, so [overflows](https://en.wikipedia.org/wiki/Integer_overflow) are quite rare. Nevertheless, if any math operation overflows, an exception will be thrown, and the transaction will fail. You could say that Tact's math is safe by default.
+
+Note, that there is no problem with mixing variables of [different state sizes](#serialization) in the same calculation. At runtime, they are all the same type no matter what — $257$-bit signed, so they all will fit.
```
Would be nice to mention the problem of integer overflow for storage. For instance, I read the value from `oneByte`, multiply the (let say non-zero) value by `256`, so there is no run-time integer overflow. And then I try to store the result back in `oneByte`. What happens next?

novusnota replied:
Failure in the action phase, right?

anton-trunov replied:
Failure in the compute phase, but exit code 5 ("Integer out of expected range")

anton-trunov replied:
We can add an explanation here and add the link to https://docs.tact-lang.org/language/ref/exit-codes. Btw, it looks like the section `5: Integer out of expected range (compute phase)` cannot be linked precisely, no anchor there.

novusnota replied:
Good call — I'll add an explanation, but won't link to exit-codes yet, due to #101 (inability to link) and #106 (re-write into the Book)

anton-trunov replied:
Created an issue so we won't forget it: https://github.com/tact-lang/tact-docs/issues/107

anton-trunov reviewed (commented):
Awesome stuff! Left a few comments

novusnota commented:
Added:
* Info on limited use of underscores (no duplicate, no trailing)
* Info on undesirable use of a leading zero in decimals

> Would be nice to mention the problem of integer overflow for storage. For instance, I read the value from `oneByte`, multiply the (let say non-zero) value by `256`, so there is no run-time integer overflow. And then I try to store the result back in `oneByte`. What happens next?

Failure in the compute phase, right? Should result in exit code 4 as per [TON Docs](https://docs.ton.org/learn/tvm-instructions/tvm-exit-codes#standard-exit-codes) AFAIK.

novusnota commented:
Added:
* Info on limited use of underscores (no duplicate, no trailing)
* Info on undesirable use of a leading zero in decimals
* An example of the compute phase error when working with integers and their serialization

Changed:
* `defining-types` is now `composite-types`, as that page also specifies how to use Structs and Messages
* Made a clean-up of the `composite-types` page
* `types` got a major clean-up and inter-linking with other parts of the Book. Additionally, added a sub-section for booleans right on it to not introduce a separate page for them. Motivation: in many languages booleans are used interchangeably with ints, so I felt that their description is closely related to the scope of this PR.
